File: 0403-frog-jump/0403-frog-jump.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int, bool>> dp;
    bool solve(vector<int>& stones, unordered_map<int, int>& mpp, int cs, int pj) {
        if (cs == stones.back()) return true;
        if (mpp.find(cs) == mpp.end() || pj <= 0) return false;
        if (dp[cs].count(pj)) return dp[cs][pj];

        return dp[cs][pj] = solve(stones, mpp, cs + pj - 1, pj - 1) ||
                            solve(stones, mpp, cs + pj, pj) ||
                            solve(stones, mpp, cs + pj + 1, pj + 1);
    }

    bool canCross(vector<int>& stones) {
        unordered_map<int, int> mpp;
        if(stones[1] != 1) return false;
        for (int s : stones) {
            mpp[s] = 1;
        }
        return solve(stones, mpp, 1, 1);
    }
};

File: 0403-frog-jump/0403-frog-jump.java
class Solution {
    Map<String, Boolean> memo = new HashMap<>();
    Map<Integer, Integer> posIndex = new HashMap<>();

    boolean solve(int[] stones, int n, int cs, int jump) {
        if (cs == stones[n - 1]) return true;
        String key = cs + "," + jump;
        if (memo.containsKey(key)) return memo.get(key);
        if (!posIndex.containsKey(cs) || jump <= 0) return false;

        boolean ans = solve(stones, n, cs + jump - 1, jump - 1)
                   || solve(stones, n, cs + jump, jump)
                   || solve(stones, n, cs + jump + 1, jump + 1);

        memo.put(key, ans);
        return ans;
    }

    public boolean canCross(int[] stones) {
        int n = stones.length;
        if (stones[1] != 1) return false;

        for (int i = 0; i < n; i++) {
            posIndex.put(stones[i], i);
        }

        return solve(stones, n, 1, 1);
    }
}
